pmlc/dialect/stripe/agginit_pass.cc

```diff
@@ -0,0 +1,239 @@
+// Copyright 2019, Intel Corporation
+
+#include "pmlc/dialect/stripe/transcode.h"
+#include "pmlc/dialect/stripe/agginit_pass.h"
+
+#include <sstream>
+#include <vector>
+
+#include "base/util/logging.h"
+
+#include "mlir/IR/Matchers.h"
+#include "mlir/Translation.h"
+
+#include "pmlc/dialect/stripe/analysis.h"
+#include "pmlc/dialect/stripe/dialect.h"
+#include "pmlc/dialect/stripe/ops.h"
+#include "pmlc/dialect/stripe/transforms.h"
+#include "pmlc/dialect/stripe/util.h"
+
+namespace pmlc {
+namespace dialect {
+namespace stripe {
+
+using vertexai::tile::codegen::proto::MLIR_AggInitPass;
+
+class AggregateInitializer {
+public:
+  AggregateInitializer(const MLIR_AggInitPass& opts, mlir::FuncOp f);
+  // Initialize the aggregate buffers in op
+  void Initialize(ParallelForOp op);
+private:
+  // Find out where to insert the initialization of tensor
+  std::pair<Operation*, Value*> InitLocation(Value* tensor);
+  // Insert the initialization for the tensor in the aggregate op
+  void InsertInit(AggregateOp aop);
+
+  // Optimization options
+  const MLIR_AggInitPass& options;
+  // The current ParallelForOp
+  ParallelForOp curOp;
+  // The current op builder
+  OpBuilder* builder;
+  // The set of tensors that are zero-initialized before this pass
+  std::set<StringRef> zeroSet;
+};
+
+AggregateInitializer::AggregateInitializer(const MLIR_AggInitPass& opts, mlir::FuncOp f)
+  : options(opts) {
+  // Collect the zero-initialized tensors
+  f.walk([this] (ZeroOp op) {
+    zeroSet.insert(tensorName(op.out()));
+  });
+  f.walk([this] (ParallelForOp op) {
+    if (hasAttr(op.getOperation(), "zero")) {
+      auto body = op.getBody();
+      for (auto& op_base : *body) {
+        if (auto ref_op = mlir::dyn_cast<RefineOp>(op_base)) {
+          zeroSet.insert(tensorName(ref_op.in()));
+        }
+      }
+    }
+  });
+}
+
+// Find out where to insert the initialization of tensor
+std::pair<Operation*, Value*> AggregateInitializer::InitLocation(Value* tensor) {
+  Value* prev_tensor = nullptr;
+  Operation* last_refine = nullptr;
+  Value* cur_tensor = tensor;
+  while (cur_tensor) {
+    if (auto def_op = cur_tensor->getDefiningOp()) {
+      if (auto allocate_op = mlir::dyn_cast<AllocateOp>(def_op)) {
+        if (!last_refine) {
```

Comment:
I think this code is not correct in the more general sense, but always works when the Stipe dialect code is generated from old stripe.  Specifically, there is no requirement of refinements on every ParallelFor, or at all even.  Let's say we have a tensor that is 1x1x1.  It's totally valid to do a 'load' directly from the allocate with no refinements, and refine(x, 0, 0, 0) might be removed by canonicalization at some point.

Reply:
For now though, I think it's not a blocker.